example_code/shmem_reduce_example.c

```diff
@@ -2,52 +2,45 @@
 #include <stdio.h>
 #include <stdlib.h>
 
-/* As if we receive some value from external source */
-long recv_a_value(unsigned seed, int npes) {
-  srand(seed);
-  return rand() % npes;
-}
-
-/* Validate the value we recieved */
-unsigned char is_valid(long value, int npes) {
-  if (value == (npes - 1))
-    return 0;
-  return 1;
-}
+#define NELEMS 32
 
 int main(void) {
   shmem_init();
   int mype = shmem_my_pe();
   int npes = shmem_n_pes();
-  size_t num = 32;
 
-  long *values = shmem_malloc(num * sizeof(long));
-  long *sums = shmem_malloc(num * sizeof(long));
+  int *values = shmem_malloc(NELEMS * sizeof(int));
 
-  unsigned char *valid_me = shmem_malloc(num * sizeof(unsigned char));
-  unsigned char *valid_all = shmem_malloc(num * sizeof(unsigned char));
+  unsigned char *value_is_maximal     = shmem_malloc(NELEMS * sizeof(unsigned char));
+  unsigned char *value_is_maximal_all = shmem_malloc(NELEMS * sizeof(unsigned char));
 
-  values[0] = recv_a_value((unsigned)mype, npes);
-  valid_me[0] = is_valid(values[0], npes);
+  static int maximal_values_count = 0;
+  static int maximal_values_total;
 
-  for (int i = 1; i < num; i++) {
-    values[i] = recv_a_value((unsigned)values[i - 1], npes);
-    valid_me[i] = is_valid(values[i], npes);
+  srand((unsigned)mype);
+
+  for (int i = 0; i < NELEMS; i++) {
+    values[i] = rand() % npes;
+
+    /* Track and count instances of maximal values (i.e., values equal to (npes-1)) */
+    value_is_maximal[i] = (values[i] == (npes - 1)) ? 1 : 0;
+    maximal_values_count += value_is_maximal[i];
   }
 
   /* Wait for all PEs to initialize reductions arrays */
   shmem_sync(SHMEM_TEAM_WORLD);
 
-  shmem_and_reduce(SHMEM_TEAM_WORLD, valid_all, valid_me, num);
-  shmem_sum_reduce(SHMEM_TEAM_WORLD, sums, values, num);
-
-  for (int i = 0; i < num; i++) {
-    if (valid_all[i]) {
-      printf("[%d] = %ld\n", i, sums[i]);
-    }
-    else {
-      printf("[%d] = invalid on one or more pe\n", i);
-    }
+  shmem_or_reduce(SHMEM_TEAM_WORLD, value_is_maximal_all, value_is_maximal, NELEMS);
+  shmem_sum_reduce(SHMEM_TEAM_WORLD, &maximal_values_total, &maximal_values_count, 1);
+
+  if (mype == 0) {
+    printf("Found %d maximal random numbers across all PEs.\n", maximal_values_total);
```

Comment:
Do we need debug statements in this example?

Reply:
Thanks for the review @naveen-rn.

I thought this was printing a "result", not a debug statement.  The result is a little arbitrary, and I was trying to keep similar semantics/output of the original program. That said, I'm totally ok with removing it...

Reply:
I think the printout helps to explain what's happening in the example. I'm ok with keeping it.